### DISCUSSION TITLE
Allow custom HTML element in htmlpurifier config

### DISCUSSION THREAD

**Discussion by thisanimus** [craftcms/cms#8808] [Ideas] [answered]
I'm using the IonIcons web component in my site, and it requires the following markup:
`<ion-icon name="checkmark-outline"></ion-icon>`
Obviously HTMLPurifier strips this out because it's not an actual HTML element.  I've added 'ion-icon' to the HTMLAllowed config string, and loaded the config, but get this error: 
`Element 'ion-icon' is not supported`
And it shows where HTMLPurifier is filtering the config options through htmlspecialchars().

My questions: 
- Is it possible to configure HTMLPurifier to allow a custom html element.
- Is this possible using the craft-supplied .json config, or do I need to write a plugin and extend HTMLPurifier?

**Comment by brandonkelly** [comment]
HTML Purifier is provided by Craft itself, so moved this over to the Craft repo.

To answer your question, it’s not currently possible, but something we intend to add support with, alongside custom attributes (#5736), so will mark this as a feature request.

**Reply by nicolasbottari** [reply]
Ran into this issue with Redactor as well. The changelogs (both Craft 3 and 4) do not mention an update for htmlpurifier to support custom HTML elements. A user is inserting malformed HTML due to turning off HTML Purify, and doing this was necessary to allow a custom (and well-formed) element in the content. Is this still not possible? Thanks!

**Comment by brandonkelly** [comment, marked as answer]
It’s possible to add a custom element via the [`EVENT_MODIFY_PURIFIER_CONFIG` event](https://github.com/craftcms/redactor#html-purifier-configs):

```php
use craft\htmlfield\events\ModifyPurifierConfigEvent;
use craft\redactor\Field;
use HTMLPurifier_Config;
use yii\base\Event;

Event::on(
    Field::class,
    Field::EVENT_MODIFY_PURIFIER_CONFIG,
    function(ModifyPurifierConfigEvent $event) {
        /** @var HTMLPurifier_Config $config */
        $config = $event->config;
        $def = $config->getDefinition('HTML', true);

        if ($def) {
            $def->addElement('element-name', /* ... */);
        }
    }
);
```

See http://htmlpurifier.org/doxygen/html/classHTMLPurifier__HTMLModule.html#affb19c170d2cdb5d1a231a73c0501895 for the `addElement()` documentation.
